namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -0,0 +1,72 @@
+package io.buoyant.namer.dnssrv
+
+import java.net.InetSocketAddress
+
+import com.twitter.finagle._
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.logging.Logger
+import com.twitter.util.Activity.State
+import com.twitter.util._
+import org.xbill.DNS
+
+class DnsSrvNamer(prefix: Path, resolver: DNS.Resolver, timer: Timer, refreshInterval: Duration, stats: StatsReceiver)
+  extends Namer {
+
+  override def lookup(path: Path): Activity[NameTree[Name]] = memoizedLookup(path)
+
+  private val log = Logger.get("dnssrv")
+  private val memoizedLookup: (Path) => Activity[NameTree[Name]] = Memoize {
+    case path@Path.Utf8(address, _) =>
+      val id = path.take(1)
+      Activity(Var.async[State[NameTree[Name]]](Activity.Pending) { state =>
+        timer.schedule(refreshInterval) {
+          val next = lookupSrv(address, prefix ++ id) match {
+            case Return(nameTree) => Activity.Ok(nameTree)
+            case Throw(e) => Activity.Failed(e)
+          }
+          state.update(next)
+        }
+      })
+    case _ => Activity.value(NameTree.Neg)
+  }
+
+  private[dnssrv] def lookupSrv(address: String, id: Path): Try[NameTree[Name]] = Try {
+    val question = DNS.Record.newRecord(
+      DNS.Name.fromString(address),
+      DNS.Type.SRV,
+      DNS.DClass.IN
+    )
+    val query = DNS.Message.newQuery(question)
+    log.debug("looking up %s", address)
+    val m = resolver.send(query)
+    log.debug("got response %s", address)
+    m.getRcode match {
+      case DNS.Rcode.NXDOMAIN =>
+        log.trace("no results for %s", address)
+        NameTree.Neg
+      case DNS.Rcode.NOERROR =>
+        val hosts = m.getSectionArray(DNS.Section.ADDITIONAL).collect {
+          case a: DNS.ARecord => a.getName -> a.getAddress
+        }.toMap
+        val srvRecords = m.getSectionArray(DNS.Section.ANSWER).collect {
+          case srv: DNS.SRVRecord =>
+            hosts.get(srv.getTarget) match {
+              case Some(inetAddress) => Address(new InetSocketAddress(inetAddress, srv.getPort))
+              case None => Address(srv.getTarget.toString, srv.getPort)
+            }
+        }
+        if (srvRecords.isEmpty) {
+          // valid DNS entry, but no instances.
+          // for some reason, NameTree.Empty doesn't work right
+          log.trace("empty response for %s", address)
+          NameTree.Neg
+        } else {
+          log.trace("got %d results for %s", srvRecords.length, address)
+          NameTree.Leaf(Name.Bound(Var.value(Addr.Bound(srvRecords: _*)), id))
+        }
+      case code =>
+        log.warning("unexpected RCODE: %s for %s", DNS.Rcode.string(code), address)
+        NameTree.Fail
```

Comment:
I think this should be a `Throw(...)` instead of a `Return(NameTree.Fail)`

Reply:
ok. I'm a little unclear on the behaviour of the different `NameTree.*` variants. When would I use `NameTree.Fail`?

Reply:
It can be useful if you want to artificially halt evaluation of a name tree, but it's rarely used in practice.